Adding retries on HTTP errors

vladbagmet opened the pull request:
### Rationale
Currently API client does not make retries on server errors. This PR changes that.

### Changes
*Describe what you changed*


[JIRA ticket](put the link to the corresponding JIRA ticket here)

#### TODO
- [ ] Update CHANGELOG.md

katsanva reviewed (changes requested):
We have removed this functionality from node SDK due to API bombing, was this change aligned with all of the teams?
@ville can you please give your input?

ville commented:
I think this should not be on by default.

vladbagmet commented:
> I think this should not be on by default.

@ville, I disabled it by default. Now users can set their own limit once api client is initialised.

oleksii-symon-corva-ai commented on src/corva/api.py:
```diff
@@ -14,6 +16,14 @@ class Api:
     """
 
     TIMEOUT_LIMITS = (3, 30)  # seconds
+    DEFAULT_MAX_RETRIES = 0
```
```suggestion
    DEFAULT_MAX_RETRIES = int(0)
```

oleksii-symon-corva-ai commented on src/corva/api.py:
```diff
@@ -31,28 +41,39 @@ def __init__(
         }
 
+    @property
+    def max_retries(self):
```
```suggestion
    def max_retries(self) -> int:
```

oleksii-symon-corva-ai commented on src/corva/api.py:
```diff
@@ -31,28 +41,39 @@ def __init__(
+        return self._max_retries
+
+    @max_retries.setter
+    def max_retries(self, value):
+        if not (isinstance(value, int) and 0 <= value <= 10):
```
```suggestion
    def max_retries(self, value: int) -> None:
        if not (0 <= value <= 10):
```

oleksii-symon-corva-ai commented on src/corva/api.py:
```diff
@@ -113,22 +135,27 @@ def _request(
-            headers=headers,
-            timeout=timeout,
-        )
+        for retry_attempt in range(max(self._max_retries, 1)):
```
If _max_retries == 1 - at most single request will be made

vladbagmet commented:
> Did you think about using popular libraries for retries like `tenacity` or `stamina`? We have single place, where we use retries - but still maybe it would lead to less logic for maintenance
> 
> Also, do we want to add this feature to docs for everyone to see?

@oleksii-symon-corva-ai, I've updated the code to make tenacity handle retries. Also, updated docs. Please review this PR again.

oleksii-symon-corva-ai commented on docs/modules/ROOT/pages/index.adoc:
```diff
@@ -269,6 +269,22 @@ include::example$api/tutorial007.py[]
+* 504
+
+[source,python]
+----
+include::example$api/tutorial008.py[]
+
```
```suggestion
----
include::example$api/tutorial008.py[]
---
```

oleksii-symon-corva-ai replied:
You can use `make docs-view` to verify docs correctness

Remember to use `branches: HEAD` here https://github.com/corva-ai/python-sdk/blob/master/docs/antora-playbook.yml#L9 before running that command

oleksii-symon-corva-ai commented on src/corva/api.py:
```diff
@@ -113,22 +171,46 @@ def _request(
+                    lambda r: r.status_code in retryable_status_codes
+                ),
+            )
+            retrying_request = retry_decorator(self._execute_request)
```
Would retrying code block here be simpler? https://tenacity.readthedocs.io/en/latest/#retrying-code-block

e.g. there would be no need to add separate `_execute_request` method

vladbagmet replied:
To be honest it was simpler before using 3-rd party retrying lib and the code we had initially was very similar to https://tenacity.readthedocs.io/en/latest/#retrying-code-block. I haven't managed make that code in example work without breaking anything.

oleksii-symon-corva-ai commented on tests/unit/test_docs/test_api.py:
```diff
@@ -120,3 +121,12 @@ def test_tutorial007(app_runner, mocker: MockerFixture):
+    time_event = ScheduledDataTimeEvent(
+        asset_id=0, company_id=0, start_time=0, end_time=0
+    )
+    time_mock = mocker.patch.object(Api, 'insert_data')
+    app_runner(tutorial008.scheduled_app, time_event)
+    time_mock.assert_not_called()
```
```suggestion
    app_runner(tutorial008.scheduled_app, time_event)
```

vladbagmet replied:
This [is addressed](https://github.com/corva-ai/python-sdk/pull/81/commits/fe5a3dfda5454d40892c4e5d386114ec1db2943c).

oleksii-symon-corva-ai reviewed (commented):
Great job, only small comments

oleksii-symon-corva-ai commented:
Also, remember to update `Unreleased` section `CHANGELOG.md` with your feature

vladbagmet commented:
> Also, remember to update `Unreleased` section in `CHANGELOG.md` with your feature

I've updated `Unreleased` section in `CHANGELOG.md` within [#fe5a3df](https://github.com/corva-ai/python-sdk/pull/81/commits/fe5a3dfda5454d40892c4e5d386114ec1db2943c).

oleksii-symon-corva-ai commented on CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+-  Added functionality to make retries for the failed HTTP requests.
```
```suggestion
## [Unreleased]

### Added
-  Functionality to make retries for the failed HTTP requests.
```

vladbagmet commented:
@oleksii-symon-corva-ai, thank you for the review!
